qcore/gsf.py

```diff
@@ -0,0 +1,210 @@
+    includes the endpoints. If length = 10, and resolution = 5, then the
+    function returns 3.
+
+      5m    5m
+    +-----+-----+
+
+    Parameters
+    ----------
+    length : float
+        Length to distribute grid points along.
+    resolution : float
+        Resolution of the grid.
+
+    Returns
+    -------
+    int
+        The number of gridpoints that fit into length.
+    """
+    return int(np.round(length / resolution + 2))
+
+
+def coordinate_meshgrid(
+    origin: np.ndarray,
+    x_upper: np.ndarray,
+    y_bottom: np.ndarray,
+    resolution: float,
+) -> np.ndarray:
+    """Creates a meshgrid of points in a bounded plane region.
+
+    Given the bounds of a rectangular planar region, create a meshgrid of
+    (lat, lon, depth) coordinates spaced at close to resolution metres apart
+    in the strike and dip directions.
+
+    origin                       x_upper
+          ┌─────────────────────┐
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │            ∧ ∧      │
+          └────────────┼─┼──────┘
+     y_bottom          └─┘
+                    resolution
+
+    Parameters
+    ----------
+    origin : np.ndarray
+        Coordinates of the origin point (lat, lon, depth).
+    x_upper : np.ndarray
+        Coordinates of the upper x boundary (lat, lon, depth).
+    y_bottom : np.ndarray
+        Coordinates of the bottom y boundary (lat, lon, depth).
+    resolution : float
+        Resolution of the meshgrid.
+
+    Returns
+    -------
+    np.ndarray
+        The meshgrid of the rectangular planar region.
+    """
+    origin = coordinates.wgs_depth_to_nztm(origin)
+    x_upper = coordinates.wgs_depth_to_nztm(x_upper)
+    y_bottom = coordinates.wgs_depth_to_nztm(y_bottom)
+    length_x = np.linalg.norm(x_upper - origin)
+    length_y = np.linalg.norm(y_bottom - origin)
+    nx = gridpoint_count_in_length(length_x, resolution)
+    ny = gridpoint_count_in_length(length_y, resolution)
+    x = np.linspace(0, length_x, nx)
+    y = np.linspace(0, length_y, ny)
+    xv, yv = np.meshgrid(x, y)
+    subdivision_coordinates = np.vstack([xv.ravel(), yv.ravel()])
+    transformation_matrix = np.vstack(
+        [(x_upper - origin) / length_x, (y_bottom - origin) / length_y]
+    ).T
+    nztm_meshgrid = (transformation_matrix @ subdivision_coordinates).T
+    nztm_meshgrid += origin
+    return coordinates.nztm_to_wgs_depth(nztm_meshgrid)
+
+
+def write_fault_to_gsf_file(
+    gsf_file_handle: TextIO,
+    meshgrids: list[np.ndarray],
+    lengths: np.ndarray,
+    widths: np.ndarray,
+    strikes: np.ndarray,
+    dips: np.ndarray,
+    rakes: np.ndarray,
+    resolution: int,
+):
```

Comment:
I don't like this function the way that it is currently. I have a `fault` module in type5 atm that encapsulates lengths, widths, strikes, dips, etc into a Fault object rather than this parallel array nonsense. Perhaps I should include this fault module into qcore? I think it has some merit as the geometry code I wrote there could be useful for other projects (like for example, the NSHMDB repo).

Reply:
If I did that, I'd also do the fault module properly, with multiple ways of instantiating faults (from corners, from centroid + strike, dip, etc). This is because the NSHMDB can read fault corners from the database, but EMOD3D specifies faults with centroids, but fundamentally they represent the same object.

Reply:
Not against having a shared Fault object.

We have in the past created a classdef file which contained this https://github.com/ucgmsim/Pre-processing/blob/f86c185c4a87ba13b3a413e78132e8ca22fa6fb3/VM/models/classdef.py#L4 (Used to also be one in empirical engine till it was removed for over complicating the problem)

Did not handle things very well but could be good to maybe bring these all together under 1 hood if implemented properly and isn't too much "overhead" just to create a fault object to get results from functions.

Reply:
This would line up with the `preprocessing` -> `source_model` repo change that we will hopefully do in the near future.